impl_core/src/main/java/io/opencensus/implcore/metrics/DerivedLongGaugeImpl.java

```diff
@@ -0,0 +1,148 @@
+
+    removeInternal(Collections.unmodifiableList(new ArrayList<LabelValue>(labelValues)));
+  }
+
+  @Override
+  public synchronized void clear() {
+    registeredPoints.clear();
+  }
+
+  @SuppressWarnings("rawtypes")
+  private synchronized void removeInternal(List<LabelValue> labelValues) {
+    Map<List<LabelValue>, PointWithFunction> registeredPointsCopy =
+        new HashMap<List<LabelValue>, PointWithFunction>(registeredPoints);
+    if (registeredPointsCopy.remove(labelValues) == null) {
+      // The element not present, no need to update the current map of time series.
+      return;
+    }
+    registeredPoints = Collections.unmodifiableMap(registeredPointsCopy);
+  }
+
+  @SuppressWarnings("rawtypes")
+  private synchronized <T> void registerTimeSeries(
+      List<LabelValue> labelValues, @Nullable T obj, ToLongFunction<T> function) {
+    PointWithFunction existingTimeSeries = registeredPoints.get(labelValues);
+    if (existingTimeSeries != null) {
+      throw new IllegalArgumentException(
+          "A different time series with the same labels already exists.");
+    }
+
+    PointWithFunction newTimeSeries = new PointWithFunction<T>(labelValues, obj, function);
+    // Updating the map of time series happens under a lock to avoid multiple add operations
+    // to happen in the same time.
+    Map<List<LabelValue>, PointWithFunction> registeredPointsCopy =
+        new HashMap<List<LabelValue>, PointWithFunction>(registeredPoints);
+    registeredPointsCopy.put(labelValues, newTimeSeries);
+    registeredPoints = Collections.unmodifiableMap(registeredPointsCopy);
+  }
+
+  @Override
+  @SuppressWarnings("rawtypes")
+  public Metric getMetric(Clock clock) {
+    Map<List<LabelValue>, PointWithFunction> currentRegisteredPoints = registeredPoints;
+    List<TimeSeries> timeSeriesList = new ArrayList<TimeSeries>(currentRegisteredPoints.size());
+    for (Map.Entry<List<LabelValue>, PointWithFunction> entry :
+        currentRegisteredPoints.entrySet()) {
+      timeSeriesList.add(entry.getValue().getTimeSeries(clock));
+    }
+    return Metric.create(metricDescriptor, timeSeriesList);
+  }
+
+  /** Implementation of {@link PointWithFunction} with a obj and function. */
+  public static final class PointWithFunction<T> {
+    private final List<LabelValue> labelValues;
+    @Nullable private final WeakReference<T> ref;
+    private final ToLongFunction<T> function;
+
+    PointWithFunction(List<LabelValue> labelValues, @Nullable T obj, ToLongFunction<T> function) {
+      this.labelValues = labelValues;
+      ref = obj != null ? new WeakReference<T>(obj) : null;
+      this.function = function;
+    }
+
+    private TimeSeries getTimeSeries(Clock clock) {
+      final T obj = ref != null ? ref.get() : null;
+
+      @SuppressWarnings("incompatible") // if obj is null
+      long value = function.applyAsLong(obj);
+
+      return TimeSeries.createWithOnePoint(
+          labelValues,
```

Comment:
Here we will re-evaluate the labelValues all the time. Consider (maybe in a separate PR) to have a helper function:

```java
class TimeSeries {
  TimeSeries setPoint(Point point) {
     return new TimeSeries(getLableValues(), Collections.singletonList(point));
  }
}
```

This way we avoid checking the labelValues all the time. Here you will have a:

```java
public static final class PointWithFunction {
  private final TimeSeries defaultTimeSeries = TimeSeries.create(labelValues);

  private TimeSeries getTimeSeries(Clock clock) {
     return defaultTimeSeries.setPoint(io.opencensus.metrics.export.Point.create(Value.longValue(value), clock.now()));
  }
}
```

Can be done separate PR. Maybe file an issue.